File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/render/VulkanContextDescriptor.java
package org.maplibre.nativeffi.render;

import java.util.Objects;

/** Mutable Vulkan backend context descriptor. */
public final class VulkanContextDescriptor {
  private NativePointer instance = NativePointer.NULL;
  private NativePointer physicalDevice = NativePointer.NULL;
  private NativePointer device = NativePointer.NULL;
  private NativePointer graphicsQueue = NativePointer.NULL;
  private int graphicsQueueFamilyIndex;
  private NativePointer getInstanceProcAddr = NativePointer.NULL;
  private NativePointer getDeviceProcAddr = NativePointer.NULL;

  public VulkanContextDescriptor() {}

  public VulkanContextDescriptor(
      NativePointer instance,
      NativePointer physicalDevice,
      NativePointer device,
      NativePointer graphicsQueue,
      int graphicsQueueFamilyIndex) {
    this.instance = Objects.requireNonNull(instance, "instance");
    this.physicalDevice = Objects.requireNonNull(physicalDevice, "physicalDevice");
    this.device = Objects.requireNonNull(device, "device");
    this.graphicsQueue = Objects.requireNonNull(graphicsQueue, "graphicsQueue");
    this.graphicsQueueFamilyIndex = graphicsQueueFamilyIndex;
  }

  public NativePointer instance() {
    return instance;
  }

  public VulkanContextDescriptor instance(NativePointer instance) {
    this.instance = Objects.requireNonNull(instance, "instance");
    return this;
  }

  public NativePointer physicalDevice() {
    return physicalDevice;
  }

  public VulkanContextDescriptor physicalDevice(NativePointer physicalDevice) {
    this.physicalDevice = Objects.requireNonNull(physicalDevice, "physicalDevice");
    return this;
  }

  public NativePointer device() {
    return device;
  }

  public VulkanContextDescriptor device(NativePointer device) {
    this.device = Objects.requireNonNull(device, "device");
    return this;
  }

  public NativePointer graphicsQueue() {
    return graphicsQueue;
  }

  public VulkanContextDescriptor graphicsQueue(NativePointer graphicsQueue) {
    this.graphicsQueue = Objects.requireNonNull(graphicsQueue, "graphicsQueue");
    return this;
  }

  public int graphicsQueueFamilyIndex() {
    return graphicsQueueFamilyIndex;
  }

  public VulkanContextDescriptor graphicsQueueFamilyIndex(int index) {
    this.graphicsQueueFamilyIndex = index;
    return this;
  }

  public NativePointer getInstanceProcAddr() {
    return getInstanceProcAddr;
  }

  public VulkanContextDescriptor getInstanceProcAddr(NativePointer getInstanceProcAddr) {
    this.getInstanceProcAddr = Objects.requireNonNull(getInstanceProcAddr, "getInstanceProcAddr");
    return this;
  }

  public NativePointer getDeviceProcAddr() {
    return getDeviceProcAddr;
  }

  public VulkanContextDescriptor getDeviceProcAddr(NativePointer getDeviceProcAddr) {
    this.getDeviceProcAddr = Objects.requireNonNull(getDeviceProcAddr, "getDeviceProcAddr");
    return this;
  }

  public VulkanContextDescriptor procAddresses(
      NativePointer getInstanceProcAddr, NativePointer getDeviceProcAddr) {
    return getInstanceProcAddr(getInstanceProcAddr).getDeviceProcAddr(getDeviceProcAddr);
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/render/VulkanContextDescriptor.kt
package org.maplibre.nativeffi.render

/** Mutable Vulkan backend context descriptor. */
public class VulkanContextDescriptor(
  instance: NativePointer,
  physicalDevice: NativePointer,
  device: NativePointer,
  graphicsQueue: NativePointer,
  graphicsQueueFamilyIndex: Int,
  getInstanceProcAddr: NativePointer,
  getDeviceProcAddr: NativePointer,
) {
  public var instance: NativePointer = instance

  public var physicalDevice: NativePointer = physicalDevice

  public var device: NativePointer = device

  public var graphicsQueue: NativePointer = graphicsQueue

  public var getInstanceProcAddr: NativePointer = getInstanceProcAddr

  public var getDeviceProcAddr: NativePointer = getDeviceProcAddr

  public var graphicsQueueFamilyIndex: Int = graphicsQueueFamilyIndex
    set(value) {
      require(value >= 0) { "graphicsQueueFamilyIndex must be non-negative" }
      field = value
    }

  init {
    require(graphicsQueueFamilyIndex >= 0) { "graphicsQueueFamilyIndex must be non-negative" }
  }
}
